### PR TITLE
fixes npn.py to get data of 3 months at a time

### PR DESCRIPTION
**Pull request by Parth-25m**

### CODE REVIEW

File: scripts/npn.py [hunk 1]
```diff
@@ -21,54 +24,139 @@ def __init__(self, **kwargs):
...
-                           ("kingdom"               ,   ("char",)       ),
-                           ("pheno_cat"             ,   ("char",)       ),
-                           ("pheno_name"            ,   ("char",)       ),
-                           ("pheno_status"          ,   ("char",)       ),
-                           ("lat"                   ,   ("double",)     ),
-                           ("lon"                   ,   ("double",)     ),
-                           ("elevation"             ,   ("int",)        ),
-                           ("network_name"          ,   ("char",)       )]
-            table.columns = columns
+
```
Comment (henrykironde) [inline scripts/npn.py]: too much spacing

**Comment by Parth-25m** [comment]
I have fixed spacing.

**Comment by henrykironde** [comment]
@Parth-25m, I think we are almost getting this dataset in and again thanks for the contribution.

I have made some suggestions but feel free to design as you think appropriate

**Comment by Parth-25m** [comment]
@henrykironde I didn't understand. Can you specify the suggestions ?

**Comment by henrykironde** [comment]
so if I ran the code with a few commented lines

```

#retriever
"""Retriever script for National Phenology Network data
The dataset contains observation data retrieved from start date to current date
date format YYYY-MM-DD
Data having a value -9999 or "-9999" are considered 'null' or 'empty
Data from the API is xml having both taxa(plantae and animalia)
"""
from future import standard_library

standard_library.install_aliases()
from builtins import str

import time
import xml.etree.ElementTree as ET

from retriever.lib.templates import Script
from retriever.lib.models import Table
from retriever import open_fr, open_fw, open_csvw, DATA_WRITE_PATH


class main(Script):
    def __init__(self, **kwargs):
        Script.__init__(self, **kwargs)
        self.name = "USA National Phenology Network"
        self.shortname = "NPN"
        self.retriever_minimum_version = '2.0.dev'
        self.version = '2.0.0'
        self.ref = "http://www.usanpn.org/results/data"
        self.tags = ["Data Type > Phenology", "Spatial Scale > Continental"]
        self.description = "The data set was collected via Nature's Notebook phenology observation program (2009-present), and (2) Lilac and honeysuckle data (1955-present)"
        self.citation = "Schwartz, M. D., Ault, T. R., & J. L. Betancourt, 2012: Spring Onset Variations and Trends in the Continental USA: Past and Regional Assessment Using Temperature-Based Indices. International Journal of Climatology (published online, DOI: 10.1002/joc.3625)."

    def download(self, engine=None, debug=False):
        Script.download(self, engine, debug)

        engine = self.engine
        csv_files = []
        request_src = "http://www.data-retriever.org/"
        base_url = "http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date={startYear}-01-01&end_date={endYear_date}&request_src={request_src}"
        start_year = 2009
        end_year = int(time.strftime("%Y"))
        header_values = ["observation_id",
                         "update_datetime",
                         "site_id",
                         "latitude",
                         "longitude",
                         "elevation_in_meters",
                         "state",
                         "species_id",
                         "genus",
                         "species",
                         "common_name",
                         "kingdom",
                         "individual_id",
                         "phenophase_id",
                         "phenophase_description",
                         "observation_date",
                         "day_of_year",
                         "phenophase_status",
                         "intensity_category_id",
                         "intensity_value",
                         "abundance_value"
                         ]

        columns = [("record_id", ("pk-auto",)),
                   ("observation_id", ("int",)),  # subsequently refered to as "status record"
                   ("update_datetime", ("char",)),
                   ("site_id", ("int",)),
                   ("latitude", ("double",)),
                   ("longitude", ("double",)),
                   ("elevation_in_meters", ("char",)),
                   ("state", ("char",)),
                   ("species_id", ("int",)),
                   ("genus", ("char",)),
                   ("species", ("char",)),
                   ("common_name", ("char",)),
                   ("kingdom", ("char",)),  # skip kingdom
                   ("individual_id", ("char",)),
                   ("phenophase_id", ("int",)),
                   ("phenophase_description", ("char",)),
                   ("observation_date", ("char",)),
                   ("day_of_year", ("char",)),
                   ("phenophase_status", ("char",)),
                   ("intensity_category_id", ("char",)),
                   ("intensity_value", ("char",)),
                   ("abundance_value", ("char",))
                   ]

        month_1 = ['1', '4', '7', '10']
        month_2 = ['3', '6', '9', '12']
        num_days = ['31', '30', '30', '31']
        index = 0
        while start_year <= end_year:
            if start_year == end_year:
                # get the data in the current year to current day (sometimes the recent day's data is empty)
                # Data starts from 01-01 to 12-31 of each year
                data_url = base_url.format(startYear=start_year, endYear_date=str(time.strftime("%Y-%d-%m")),
                                           request_src=request_src)
            else:
                startYear = str(start_year) + "-" + month_1[index] + "-1"
                endYear_date = str(start_year) + "-" + month_2[index] + '-' + num_days[index]
                data_url = base_url.format(startYear=startYear, endYear_date=endYear_date,
                                           request_src=request_src)

            xml_file_name = '{}'.format(startYear) + ".xml"

            print(data_url)
        #     engine.download_file(data_url, xml_file_name)
        #
        #     # Create csv files for that year
        #     csv_observation = '{}'.format(startYear) + ".csv"
        #     csv_files.append(csv_observation)
        #     csv_buff = open_fw(engine.format_filename(csv_observation))
        #     csv_writer = open_csvw(csv_buff)
        #
        #     csv_writer.writerow(header_values)
        #
        #     ''' Parse xml to read data
        #     file_output = open_fr(engine.format_filename(xml_file_name), encode=False)
        #     v = file_output.read()'''
        #
        #     file_read = ""
        #     fname = DATA_WRITE_PATH.strip('{dataset}') + 'NPN/' + xml_file_name
        #     with open(fname, 'r') as fp1:
        #       file_read = fp1.read()
        #
        #     root = ET.fromstring(file_read)
        #
        #     for elements in root:
        #         index_map = {val: i for i, val in enumerate(header_values)}
        #         diction = sorted(elements.attrib.items(), key=lambda pair: index_map[pair[0]])
        #         csv_writer.writerow([x[1] for x in diction])
        #
        #     csv_buff.close()
        #
            index = (index + 1) % 4
            if(index == 0):
              start_year = start_year + 1
        #
        # # Create table
        # table = Table('obsercations', delimiter=',', pk='record_id', contains_pk=True)
        # table.columns = columns
        # engine.table = table
        # engine.create_table()
        # for data_file in csv_files:
        #     engine.insert_data_from_file(engine.find_file(data_file))
        return engine

SCRIPT = main()


```
I get 

```
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2009-1-1-01-01&end_date=2009-3-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2009-4-1-01-01&end_date=2009-6-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2009-7-1-01-01&end_date=2009-9-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2009-10-1-01-01&end_date=2009-12-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2010-1-1-01-01&end_date=2010-3-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2010-4-1-01-01&end_date=2010-6-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2010-7-1-01-01&end_date=2010-9-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2010-10-1-01-01&end_date=2010-12-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2011-1-1-01-01&end_date=2011-3-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2011-4-1-01-01&end_date=2011-6-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2011-7-1-01-01&end_date=2011-9-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2011-10-1-01-01&end_date=2011-12-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2012-1-1-01-01&end_date=2012-3-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2012-4-1-01-01&end_date=2012-6-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2012-7-1-01-01&end_date=2012-9-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2012-10-1-01-01&end_date=2012-12-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2013-1-1-01-01&end_date=2013-3-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2013-4-1-01-01&end_date=2013-6-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2013-7-1-01-01&end_date=2013-9-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2013-10-1-01-01&end_date=2013-12-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2014-1-1-01-01&end_date=2014-3-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2014-4-1-01-01&end_date=2014-6-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2014-7-1-01-01&end_date=2014-9-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2014-10-1-01-01&end_date=2014-12-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2015-1-1-01-01&end_date=2015-3-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2015-4-1-01-01&end_date=2015-6-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2015-7-1-01-01&end_date=2015-9-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2015-10-1-01-01&end_date=2015-12-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2016-1-1-01-01&end_date=2016-3-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2016-4-1-01-01&end_date=2016-6-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2016-7-1-01-01&end_date=2016-9-30&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2016-10-1-01-01&end_date=2016-12-31&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2017-01-01&end_date=2017-29-01&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2017-01-01&end_date=2017-29-01&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2017-01-01&end_date=2017-29-01&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2017-01-01&end_date=2017-29-01&request_src=http://www.data-retriever.org/
```


**urls for the last year**
These look like they are fetching the same url many times

```
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2017-01-01&end_date=2017-29-01&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2017-01-01&end_date=2017-29-01&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2017-01-01&end_date=2017-29-01&request_src=http://www.data-retriever.org/
http://www.usanpn.org/npn_portal/observations/getObservations.xml?start_date=2017-01-01&end_date=2017-29-01&request_src=http://www.data-retriever.org/
```

**Comment by henrykironde** [comment]
Also try using pycharm, to catch most of the pep clean ups needed, like spaces after or before operators and etc

**Comment by henrykironde** [comment]
Hi @Parth-25m, this does not seem to pass on python3
refer to http://stackoverflow.com/questions/5973120/what-is-the-replacement-for-dateutil-parser-in-python3

**Comment by henrykironde** [comment]
lets remove the print statements` print(data_url + "\n") `

**Comment by Parth-25m** [comment]
@henrykironde 
I used an extra python module - dateutil
You need to install it in python3 for npn.py to work.

**Comment by Parth-25m** [comment]
@henrykironde I will do it without using extra module.

**Comment by henrykironde** [comment]
@Parth-25m thanks, I will test the PR tomorrow.

**Comment by henrykironde** [comment]
@Parth-25m with these last requested edits, this will be ready to get ,merged.

We need to overwrite this entire commit history so that we keep one commit. 
It will be good if you update the branch to the current weecology master. 
If you have difficulty in squashing the commits let me know. 

Also change the version number of this script version txt

**Comment by Parth-25m** [comment]
@henrykironde I made a mistake while squashing the commits. It led to eventually closing this branch. 
Could you tell me what should I put as the updated version number of ```npn.py```. I will open a new pull request with the required changes. Thanks!

**Comment by henrykironde** [comment]
No worries, get a your master updated, checkout a new branch and then add the changes.

**Comment by henrykironde** [comment]
If you don't have changes i could get you the copy you made.

**Pull request closed without merging**